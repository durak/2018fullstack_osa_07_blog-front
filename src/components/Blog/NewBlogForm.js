import PropTypes from 'prop-types'
import React from 'react'
import { connect } from 'react-redux'
import { Redirect } from 'react-router-dom'
import { Form, Button, Input } from 'semantic-ui-react'

import { blogCreate } from '../../reducers/blogReducer'
import { notify } from '../../reducers/notificationReducer'

class NewBlogForm extends React.Component {
  static propTypes = {
    blogCreate: PropTypes.func.isRequired,
    notify: PropTypes.func.isRequired
  }

  constructor(props) {
    super(props)
    this.state = {
      visible: false,
      title: '',
      author: '',
      url: '',
      valid: {
        title: true,
        author: true,
        url: true
      },
      redirectAfterSubmit: null
    }
  }

  componentDidUpdate(prevProps, prevState) {
    if (prevState.redirectAfterSubmit !== this.state.redirectAfterSubmit) {
      this.setState({ redirectAfterSubmit: null })
    }
  }

  validateInputs = () => {
    const valid = {
      title: this.state.title.length > 0,
      author: this.state.author.length > 0,
      url: this.state.url.length > 0
    }

    this.setState({
      valid
    })

    return valid.title && valid.author && valid.url
  }

  handleChange =  (event) => {
    this.setState({
      [event.target.name]: event.target.value
    })
  }

  handleSubmit = async (e) => {
    e.preventDefault()
    if (!await this.validateInputs()) return

    const newBlog = {
      title: this.state.title,
      author: this.state.author,
      url: this.state.url
    }

    const response = await this.props.blogCreate(newBlog)
    this.props.notify(`a new blog ${newBlog.title} by ${newBlog.author} added`, 'message')
    this.props.togglable.current.toggleVisibility()
    console.log('RESPOEEEEENNENSNSNNENE', response)

    this.setState({
      title: '', author: '', url: '',
      valid: { title: true, author: true, url: true },
      redirectAfterSubmit: response.id
    })
  }


  render() {

    if (this.state.redirectAfterSubmit) {
      return(<Redirect push to={`/blogs/${this.state.redirectAfterSubmit}`} />)
    }


    return (

      <Form onSubmit={this.handleSubmit}>
        <Form.Group widths="equal">
          <Form.Field>
            <label>title</label>
            <Form.Input
              fluid
              type="text"
              name="title"
              value={this.state.title}
              onChange={this.handleChange}
              error={!this.state.valid.title}
            />
          </Form.Field>

          <Form.Field>
            <label>author</label>
            <Form.Input
              fluid
              type="text"
              name="author"
              value={this.state.author}
              onChange={this.handleChange}
              error={!this.state.valid.author}
            />
          </Form.Field>

          <Form.Field>
            <label>url</label>
            <Form.Input
              fluid
              type="text"
              name="url"
              value={this.state.url}
              onChange={this.handleChange}
              error={!this.state.valid.url}
            />
          </Form.Field>

          <Button type="submit">submit</Button>
        </Form.Group>
      </Form>

    )
  }
}

export default connect(
  null,
  { blogCreate, notify }
)(NewBlogForm)