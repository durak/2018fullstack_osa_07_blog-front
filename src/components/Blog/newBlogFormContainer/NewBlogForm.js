import React from 'react'
import PropTypes from 'prop-types'
import { Form, Button } from 'semantic-ui-react'

const NewBlogForm = ({
  handleSubmit, handleChange, handleCancel,
  title, author, url, valid
}) => {

  return (

    <Form onSubmit={handleSubmit}>
      <h1>Add new blog</h1>
      <Form.Group widths="equal">
        <Form.Field>
          <label>title</label>
          <Form.Input
            fluid
            type="text"
            name="title"
            value={title}
            onChange={handleChange}
            error={!valid.title}
          />
        </Form.Field>

        <Form.Field>
          <label>author</label>
          <Form.Input
            fluid
            type="text"
            name="author"
            value={author}
            onChange={handleChange}
            error={!valid.author}
          />
        </Form.Field>

        <Form.Field>
          <label>url</label>
          <Form.Input
            fluid
            type="text"
            name="url"
            value={url}
            onChange={handleChange}
            error={!valid.url}
          />
        </Form.Field>


      </Form.Group>
      <Button type="submit">submit</Button>
      <Button negative type="reset" onClick={handleCancel}>cancel</Button>

    </Form>
  )
}

export default NewBlogForm

NewBlogForm.propTypes = {
  handleSubmit: PropTypes.func.isRequired,
  handleChange: PropTypes.func.isRequired,
  handleCancel: PropTypes.func.isRequired,
  title: PropTypes.string.isRequired,
  author: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
  valid: PropTypes.object.isRequired
}