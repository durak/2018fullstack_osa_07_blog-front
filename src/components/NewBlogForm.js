import PropTypes from 'prop-types'
import React from 'react'

class NewBlogForm extends React.Component {
  static propTypes = {
    addBlog: PropTypes.func.isRequired    
  }

  constructor(props) {
    super(props)
    this.state = {
      visible: false,
      title: '',
      author: '',
      url: ''
    }
  }

  handleChange = (event) => {
    this.setState({
      [event.target.name]: event.target.value
    })
  }

  handleSubmit = (e) => {
    e.preventDefault()

    const newBlog = {
      title: this.state.title,
      author: this.state.author,
      url: this.state.url
    }

    this.setState({
      title: '', author: '', url: ''
    })

    this.props.addBlog(newBlog)
  }


  render() {


    return (



      <div>
        <form onSubmit={this.handleSubmit}>
          <div>
            title
              <input
              type="text"
              name="title"
              value={this.state.title}
              onChange={this.handleChange}
            />
          </div>
          <div>
            author
              <input
              type="text"
              name="author"
              value={this.state.author}
              onChange={this.handleChange}
            />
          </div>
          <div>
            url
              <input
              type="text"
              name="url"
              value={this.state.url}
              onChange={this.handleChange}
            />
          </div>
          <button type="submit">lisää</button>
        </form>
      </div>

    )
  }
}

export default NewBlogForm