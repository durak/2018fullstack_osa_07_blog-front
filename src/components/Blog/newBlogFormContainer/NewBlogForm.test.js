import React from 'react'
import { shallow, mount } from 'enzyme'
import renderer from 'react-test-renderer'
import NewBlogForm from './NewBlogForm'

describe('<NewBlogForm />', () => {

  const mockHandleSubmit = jest.fn()
  const mockHandleCancel = jest.fn()
  const mockHandleChange = jest.fn()


  it('renders correctly', () => {
    const tree = renderer.create(
      <NewBlogForm
        handleSubmit={mockHandleSubmit}
        handleChange={mockHandleChange}
        handleCancel={mockHandleCancel}
        title='has title'
        author='has author'
        url='has url'
        valid={{ title: true, author: true, url: true }}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot()
  })

  it('calls handlers when submit or cancel is clicked or value changes in input fields', () => {
    const Wrapper = (props) => {
      return (
        <NewBlogForm
          handleSubmit={props.handleSubmit}
          handleChange={props.handleChange}
          handleCancel={props.handleCancel}
          title='has title'
          author='has author'
          url='has url'
          valid={{ title: true, author: true, url: true }}
        />
      )
    }

    const component = mount(
      <Wrapper
        handleSubmit={mockHandleSubmit}
        handleChange={mockHandleChange}
        handleCancel={mockHandleCancel}
      />
    )

    const buttons = component.find('Button')
    buttons.at(0).simulate('submit')
    expect(mockHandleSubmit.mock.calls.length).toBe(1)
    buttons.at(1).simulate('click')
    expect(mockHandleCancel.mock.calls.length).toBe(1)

    const input = component.find('input')
    input.at(0).simulate('change', { target: { value: 'asd' } })
    input.at(1).simulate('change', { target: { value: 'asd' } })
    input.at(2).simulate('change', { target: { value: 'asd' } })
    expect(mockHandleChange.mock.calls.length).toBe(3)
  })
})


