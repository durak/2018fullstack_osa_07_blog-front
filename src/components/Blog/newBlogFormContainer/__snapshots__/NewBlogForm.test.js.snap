// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewBlogForm /> renders correctly 1`] = `
<form
  action={undefined}
  className="ui form"
  onSubmit={[Function]}
>
  <h1>
    Add new blog
  </h1>
  <div
    className="equal width fields"
  >
    <div
      className="field"
    >
      <label>
        title
      </label>
      <div
        className="field"
      >
        <div
          className="ui fluid input"
          content={undefined}
        >
          <input
            disabled={undefined}
            name="title"
            onChange={[Function]}
            required={undefined}
            tabIndex={undefined}
            type="text"
            value="has title"
          />
        </div>
      </div>
    </div>
    <div
      className="field"
    >
      <label>
        author
      </label>
      <div
        className="field"
      >
        <div
          className="ui fluid input"
          content={undefined}
        >
          <input
            disabled={undefined}
            name="author"
            onChange={[Function]}
            required={undefined}
            tabIndex={undefined}
            type="text"
            value="has author"
          />
        </div>
      </div>
    </div>
    <div
      className="field"
    >
      <label>
        url
      </label>
      <div
        className="field"
      >
        <div
          className="ui fluid input"
          content={undefined}
        >
          <input
            disabled={undefined}
            name="url"
            onChange={[Function]}
            required={undefined}
            tabIndex={undefined}
            type="text"
            value="has url"
          />
        </div>
      </div>
    </div>
  </div>
  <button
    className="ui button"
    disabled={undefined}
    onClick={[Function]}
    role="button"
    tabIndex={undefined}
    type="submit"
  >
    submit
  </button>
  <button
    className="ui negative button"
    disabled={undefined}
    onClick={[Function]}
    role="button"
    tabIndex={undefined}
    type="reset"
  >
    cancel
  </button>
</form>
`;
